UPGRADE.md

```diff
@@ -0,0 +1,119 @@
+# Upgrade Guide
+
+This document provides guidelines for upgrading between versions of Flow PHP. 
+Please follow the instructions for your specific version to ensure a smooth upgrade process.
+
+---
+
+## Upgrading from 0.3.x to 0.4.x
+
+### 1) `ref` expression nullability 
+
+`ref("entry_name")` is no longer returning null when the entry is not found. Instead, it throws an exception.
+The same behavior can be achieved through using a newly introduced `optional` expression: 
+
+Before:
+```php
+<?php
+
+use function Flow\ETL\DSL\optional;
+use function Flow\ETL\DSL\ref;
+
+ref('non_existing_column')->cast('string'); 
+```
+
+After: 
+```php
+<?php
+
+use function Flow\ETL\DSL\optional;
+use function Flow\ETL\DSL\ref;
+
+optional(ref('non_existing_column'))->cast('string');
+// or  
+optional(ref('non_existing_column')->cast('string'));
+```
+
+### 2) Extractors output
```

Comment:
There should be a point about extractor construct change that removes the row name in favor of the entry factory.